### PR TITLE
(PDB-1014) Make factsets hash field not nullable

### PR DESCRIPTION
**Pull request by senior**

### CODE REVIEW

**Comment by senior** [comment]
@kbarber is this what you had in mind for [PDB-1014](https://tickets.puppetlabs.com/browse/PDB-1014)? It seems like this could have been done when it was first introduced, so I'm not entirely sure what you had in mind.

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 1]
```diff
@@ -1465,6 +1465,12 @@
       (format "ALTER TABLE reports ADD COLUMN logs %s DEFAULT NULL" jsonb-type)
       (format "ALTER TABLE reports ADD COLUMN resources %s DEFAULT NULL" jsonb-type))))
 
+(defn factset-hash-field-not-nullable
+  []
+  (jdbc/do-commands
+   "UPDATE factsets SET hash='' WHERE hash is NULL"
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: is this going to do anything weird with the existing unique constraint on hash?

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: yeah, if you have null-valued hashes this migration causes a constraint violation.

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: just to echo what I said in the channel the other day, I think a simple way to make this work would be to set the hash to the stringified factset id.

Comment (kbarber) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: So I'm trying to figure out the idea here, behind why the ticket was raised.

I can't help but think that setting this to anything other than a real hash will cause problems with HA for older data that doesn't have one already at migration time. I mean, if you send this content over via HA, won't it generate a different hash for comparison? /cc @mullr

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: I don't think we're using hashes for that, otherwise we'd have problems when we change hash fields too.

Comment (senior) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: I think we'd be fine with a pk as a string for that value. It'll be updated on next facset anyway. Why I am asking is we could have always done that, there's nothing 4.0.0 specific about that.

Comment (kbarber) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: I'm pretty sure our thinking was to generate a hash later on, due to HA concerns. If this is not a thing, then yeah it didn't/doesn't matter.

**Comment by senior** [comment]
@wkalt fixed

**Comment by wkalt** [comment]
looks good to me, just waiting for tests

**Pull request merged**